allennlp_models/seq2seq/bart/bart.py

```diff
@@ -0,0 +1,401 @@
+        # to not mess up the positional embeddings in the decoder.
+        padding_size = 0
+        if step > 0:
+            padding_size = step + 1
+            padding = torch.full(
+                (last_predictions.shape[0], padding_size),
+                self._pad_id,
+                dtype=last_predictions.dtype,
+                device=last_predictions.device,
+            )
+            last_predictions = torch.cat([padding, last_predictions], dim=-1)
+
+        decoder_cache = None
+        decoder_cache_dict = {
+            k: (state[k].contiguous() if state[k] is not None else None)
+            for k in state
+            if k not in {"input_ids", "input_mask", "encoder_states"}
+        }
+        if len(decoder_cache_dict) != 0:
+            decoder_cache = self._dict_to_decoder_cache(decoder_cache_dict)
+
+        decoder_logits = None
+        for i in range(padding_size, last_predictions.shape[1]):
+            encoder_outputs = (
+                (state["encoder_states"],) if state["encoder_states"] is not None else None
+            )
+            outputs = self.bart(
+                input_ids=state["input_ids"],
+                attention_mask=state["input_mask"],
+                encoder_outputs=encoder_outputs,
+                decoder_input_ids=last_predictions[:, : i + 1],
+                decoder_cached_states=decoder_cache,
+                generation_mode=True,
+            )
+            if decoder_logits is None:
+                decoder_logits = outputs[0]
+            else:
+                idx = last_predictions[:, i].view(-1, 1, 1)
+                decoder_logits = outputs[0] + decoder_logits.gather(dim=-1, index=idx)
+
+            if self.bart.config.output_past:
+                decoder_cache = outputs[1][1]
+
+            state["encoder_states"] = outputs[2]
+
+        if decoder_cache is not None:
+            decoder_cache_dict = self._decoder_cache_to_dict(decoder_cache)
+            state.update(decoder_cache_dict)
+
+        return decoder_logits[:, 0], state
+
+    @overrides
+    def make_output_human_readable(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, Any]:
+        """
+
+        # Parameters
+
+        output_dict : `Dict[str, torch.Tensor]`
+            A dictionary containing a batch of predictions with key `predictions`. The tensor should have
+            shape `(batch_size, max_sequence_length)`
+
+        # Returns
+
+        Dict[str, Any]
+            Original `output_dict` with an additional `predicted_tokens` key that maps to a list of lists of
+            tokens.
+
+        """
+        predictions = output_dict["predictions"]
+        predicted_tokens = [None] * predictions.shape[0]
+        for i in range(predictions.shape[0]):
+            predicted_tokens[i] = self._indexer.indices_to_tokens(
+                {"token_ids": predictions[0].tolist()}, self.vocab
+            )
+        output_dict["predicted_tokens"] = predicted_tokens
```

Comment:
Do these functions usually modify their inputs? If so, carry on. If not, I'd prefer it if this one also didn't.

Reply:
Yes, in fact this is explicitly noted in the comment of the base method:

`This method `modifies` the input dictionary, and also `returns` the same dictionary.`